Userland/Utilities/hexdump.cpp

```diff
@@ -30,44 +32,52 @@ int main(int argc, char** argv)
         file = file_or_error.value();
     }
 
-    auto contents = file->read_all();
-
-    Vector<u8, 16> line;
-
-    auto print_line = [&] {
-        for (size_t i = 0; i < 16; ++i) {
-            if (i < line.size())
-                out("{:02x} ", line[i]);
+    auto print_line = [](u8* buf, size_t size) {
+        VERIFY(size <= LINE_LENGTH_BYTES);
+        for (size_t i = 0; i < LINE_LENGTH_BYTES; ++i) {
+            if (i < size)
+                out("{:02x} ", buf[i]);
             else
                 out("   ");
 
             if (i == 7)
                 out("  ");
         }
 
-        out("  ");
+        out("  |");
 
-        for (size_t i = 0; i < 16; ++i) {
-            if (i < line.size() && isprint(line[i]))
-                putchar(line[i]);
+        for (size_t i = 0; i < size; ++i) {
+            if (isprint(buf[i]))
+                putchar(buf[i]);
             else
-                putchar(' ');
+                putchar('.');
         }
 
+        putchar('|');
         putchar('\n');
     };
 
-    for (size_t i = 0; i < contents.size(); ++i) {
-        line.append(contents[i]);
-
-        if (line.size() == 16) {
-            print_line();
-            line.clear();
+    Array<u8, BUFSIZ> contents;
+    static_assert(LINE_LENGTH_BYTES * 2 <= contents.size(), "Buffer is too small?!");
+    size_t contents_size = 0;
+
+    int nread;
+    do {
+        nread = file->read(&contents[contents_size], BUFSIZ - contents_size);
+        contents_size += nread;
+        // Print as many complete lines as we can (possibly none).
+        size_t offset;
+        for (offset = 0; offset + LINE_LENGTH_BYTES - 1 < contents_size; offset += LINE_LENGTH_BYTES) {
+            print_line(&contents[offset], LINE_LENGTH_BYTES);
         }
-    }
-
-    if (!line.is_empty())
-        print_line();
+        for (size_t i = 0; offset + i < contents_size; ++i) {
+            contents[i] = contents[offset + i];
+        }
+        contents_size -= offset;
+    } while (nread);
+    VERIFY(contents_size <= LINE_LENGTH_BYTES - 1);
+    if (contents_size)
```

Comment:
Might be more `readable` ( ;) ) to explicitly check `if (contents_size > 0)` here.